crypto/keys/ecdsa/ecdsa_pubkey.go

```diff
@@ -0,0 +1,166 @@
+
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	"github.com/cosmos/cosmos-sdk/types/address"
+	"github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+// signature holds the r and s values of an ECDSA signature
+type signature struct {
+	R, S *big.Int
+}
+
+type ecdsaPK struct {
+	ecdsa.PublicKey
+
+	// cache
+	address tmcrypto.Address
+}
+
+// String implements PubKey interface
+func (pk *ecdsaPK) Address() tmcrypto.Address {
+	if pk.address == nil {
+		pk.address = address.Hash(curveNames[pk.Curve], pk.Bytes())
+	}
+	return pk.address
+}
+
+// String implements PubKey interface
+func (pk *ecdsaPK) String() string {
+	return fmt.Sprintf("%s{%X}", curveNames[pk.Curve], pk.Bytes())
+}
+
+// Bytes returns the byte representation of the public key using a compressed form
+// specified in section 4.3.6 of ANSI X9.62 with first byte being the curve type.
+func (pk *ecdsaPK) Bytes() []byte {
+	if pk == nil {
+		return nil
+	}
+	compressed := make([]byte, PubKeySize)
+	compressed[0] = curveTypes[pk.Curve]
+	compressed[1] = byte(pk.Y.Bit(0)) | 2
+	pk.X.FillBytes(compressed[2:])
+	return compressed
+}
+
+// Equals - you probably don't need to use this.
+// Runs in constant time based on length of the keys.
+func (pk *ecdsaPK) Equals(other cryptotypes.PubKey) bool {
+	pk2, ok := other.(*ecdsaPK)
+	if !ok {
+		return false
+	}
+
+	return pk.PublicKey.Equal(&pk2.PublicKey)
+}
+
+// VerifySignature implements skd.PubKey interface
+func (pk *ecdsaPK) VerifySignature(msg []byte, sig []byte) bool {
+	s := new(signature)
+	if _, err := asn1.Unmarshal(sig, s); err != nil || s == nil {
+		return false
+	}
+
+	h := sha256.Sum256(msg)
+	return ecdsa.Verify(&pk.PublicKey, h[:], s.R, s.S)
+}
+
+// Type returns key type name. Implements sdk.PubKey interface
+func (pk *ecdsaPK) Type() string {
+	return curveNames[pk.Curve]
+}
+
+// **** proto.Message ****
+
+func (pk *ecdsaPK) Reset() {
+	*pk = ecdsaPK{}
+}
+func (*ecdsaPK) ProtoMessage() {}
+
+// **** Proto Marshaler ****
+
+// Marshal implements ProtoMarshaler interface
+func (pk *ecdsaPK) Marshal() ([]byte, error) {
+	bv := gogotypes.BytesValue{Value: pk.Bytes()}
+	return bv.Marshal()
+}
+
+// MarshalTo implements ProtoMarshaler interface
+func (pk *ecdsaPK) MarshalTo(data []byte) (int, error) {
+	bv := gogotypes.BytesValue{Value: pk.Bytes()}
+	return bv.MarshalTo(data)
+}
+
+// MarshalToSizedBuffer implements ProtoMarshaler interface
+func (pk *ecdsaPK) MarshalToSizedBuffer(dAtA []byte) (int, error) {
+	bv := gogotypes.BytesValue{Value: pk.Bytes()}
+	return bv.MarshalToSizedBuffer(dAtA)
+}
+
+// Unmarshal implements ProtoMarshaler interface
+func (pk *ecdsaPK) Unmarshal(b []byte) error {
+	bv := gogotypes.BytesValue{}
+	err := bv.Unmarshal(b)
+	if err != nil {
+		return err
+	}
+	return pk.UnmarshalAmino(bv.Value)
+}
+
+// Size implements ProtoMarshaler interface.
+func (pk *ecdsaPK) Size() int {
+	if pk == nil {
+		return 0
+	}
+	return sovPubKeySize
+}
+
+// **** Amino Marshaler ****
```

Comment:
Why?

Reply:
may be good to remove this. Is there a need to support amino with new keys?

Reply:
I suspect it's because of multisig perhaps? But if that is the only reason, I think we may not want to support multisig with the new keys.

Reply:
> should docs be added around this key type. the known vulnerabilities should be added as a warning as well.

@marbar3778  - I added more docs. To my knowledge there are no known vulnerabilities. If you are aware about any then lmk please.

Reply:
I think you've added a comment to the wrong thread Robert. We're asking about amino support for the new keys?

Reply:
hmm, strange - I clicked "Quote replay" and github posted it here.

Reply:
As for Amino - all our keys are are supporting it, so for compatibility I added it.

Reply:
Either way, there is no need to implement any of these methods. Amino knows how to marshal regular structs. All that would be needed is a codec registration. Please delete all this.

Reply:
amino not used any more